### PR TITLE
Implement new zarr upload process

### LINKED ISSUE #1439: Implement Zarr upload process simplification spec
[Design doc](https://github.com/dandi/dandi-archive/blob/master/doc/design/zarr-performance-redesign.md)

### PR DESCRIPTION
**Pull request by AlmightyYakob**
Closes #1439

This both implements the new process for zarr upload, as well as removes a lot of now unused code.

**API Changes**:
* The concept of synchronous upload, and upload batches, is gone. As a consequence, the following endpoints have been removed:
	* `POST /zarr/{zarr_id}/upload/`
	* `DELETE /zarr/{zarr_id}/upload/`
	* `POST /zarr/{zarr_id}/upload_complete/`
* To request presigned upload URLs, the new `POST /zarr/{zarr_id}/files/` endpoint should be used. The payload for this endpoint is just a list of paths (strings), instead of a the previous implementation, which was a list of objects. It also simply returns a list of presigned URLs, corresponding in order to the provided list of paths. 
* The `POST /zarr/{zarr_id}/ingest/` endpoint has been replaced with the `POST /zarr/{zarr_id}/finalize/` endpoint.

### CODE REVIEW

**Comment by jwodder** [comment]
@AlmightyYakob Are the `.../files` and `.../finalize` endpoints supposed to end in a slash?  The endpoints in this API normally do so, but you've left that off here.

**Comment by jwodder** [comment]
@AlmightyYakob I've prepared a PR to update the client for this [here](https://github.com/dandi/dandi-cli/pull/1202).  However, when I tested the code against a local Docker image of this branch, when it got to [this test](https://github.com/dandi/dandi-cli/blob/e81be32aaf79654367d1cac20e9a262832678fd2/dandi/tests/test_upload.py#L286) (which uploads a Zarr, then uploads a different set of entries to the Zarr necessitating deleting & replacing all entries from the first upload), the upload got stuck for several minutes waiting for the Zarr to leave the "Ingesting" state, and I eventually gave up and killed the tests.  Querying the local API at the `/files/` endpoint while this was happening correctly returned the new set of Zarr entries.

**Comment by AlmightyYakob** [comment]
> @AlmightyYakob Are the `.../files` and `.../finalize` endpoints supposed to end in a slash? The endpoints in this API normally do so, but you've left that off here.

Correct, I've fixed it.

**Comment by jwodder** [comment]
@AlmightyYakob Testing locally with the latest version of this branch still results in the same test taking more than two minutes to ingest a very small Zarr.  Moreover, among all the Zarr tests in [`test_upload.py`](https://github.com/dandi/dandi-cli/blob/e81be32aaf79654367d1cac20e9a262832678fd2/dandi/tests/test_upload.py), the following take more than two minutes to ingest:

- `test_upload_different_zarr`
- `test_upload_loose_zarr`
- `test_upload_different_zarr_entry_conflicts`
- `test_upload_different_zarr_file_to_parent_dir`

The affected tests are those which create & upload to a Zarr, ingest it, then modify it and (try to) ingest it again.

**Comment by AlmightyYakob** [comment]
> @AlmightyYakob Testing locally with the latest version of this branch still results in the same test taking more than two minutes to ingest a very small Zarr. Moreover, among all the Zarr tests in [`test_upload.py`](https://github.com/dandi/dandi-cli/blob/e81be32aaf79654367d1cac20e9a262832678fd2/dandi/tests/test_upload.py), the following take more than two minutes to ingest:
> 
> * `test_upload_different_zarr`
> * `test_upload_loose_zarr`
> * `test_upload_different_zarr_entry_conflicts`
> * `test_upload_different_zarr_file_to_parent_dir`
> 
> The affected tests are those which create & upload to a Zarr, ingest it, then modify it and (try to) ingest it again.

Thanks for the info, I'll look into it. In the meantime, I'll put this pull request back into a draft state.

**Comment by jwodder** [comment]
@AlmightyYakob The CLI tests pass with the latest changes to this PR.

**Review by waxlamp — APPROVED** [review]

File: dandiapi/zarr/tests/utils.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import hashlib
+import os
+from pathlib import Path
+
+import faker
+from zarr_checksum.generators import ZarrArchiveFile
```
Comment (danlamanna) [inline dandiapi/zarr/tests/utils.py:6]: This is sort of a criticism of `zarr-checksum`, but having `ZarrArchive` and `ZarrArchiveFile` interleaved in code is confusing. I'm bad at names, but `ZarrFileDescriptor` or something might be better.

**Review by danlamanna — APPROVED** [review]
LGTM, left a comment but it's nothing that needs to be resolved before merging.

**Comment by AlmightyYakob** [comment]
I tested the CLI integration by using the [`archive-1448`](https://github.com/dandi/dandi-cli/pull/1202) branch, which is the one made to work with this change, instead of `master`: https://github.com/dandi/dandi-archive/actions/runs/4147238904/jobs/7173976149

It passed, and I reverted the commit to point back to master.

**Pull request merged**